### DISCUSSION TITLE
Spring Test에서 @Mock과 @Spy의 차이는 무엇인가요?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#168] [Spring Test]
.

**Comment by 3Juhwan** [comment]
@MockBean @SpyBean 도 다뤄주세요~~

**Comment by J-I-H-O** [comment]
**<`@Mock` vs `@Spy`>**

`@Mock` 어노테이션은 모의 객체를 생성합니다. 
**실제 객체의 메서드 호출은 일어나지 않습니다.** 기본적으로 **전체 메서드가 모의**되어 반환형에 따라 **기본값을 반환**합니다. 이때 사전에 정의해둔 동작을 수행하도록 할 수 있습니다.
Mock 객체는 **상태를 가지지 않는다는 특징**이 있습니다.

이와 달리 `@Spy` 어노테이션은 실제 객체를 스파이로 감싸서 생성합니다. 
스파이 객체는 **실제 객체의 메서드를 호출**할 수 있으며, 필요에 따라 **특정 메서드만 모의**할 수 있습니다.
스파이 객체는 **실제 객체의 상태를 유지**합니다.

따라서 `@Mock` 은 외부 의존성을 완전히 격리하고 싶을 때 사용하고, `@Spy` 는 실제 객체의 일부 메서드만 mocking하고 나머지 메서드는 실제 동작을 수행하도록 하고 싶을 때 사용할 수 있습니다.

---

**<`@Mock` 과 `@MockBean`>**

> `@Mock` : `org.mockito`
`@MockBean` : `org.springframework.boot.test.mock.mockito.MockBean`
> 

Spring 없이도 동작하는 `@Mock` 과는 달리, `@MockBean` 은 SpringBootTest에서 사용되는 어노테이션입니다.
애플리케이션 **컨텍스트에서 관리되는 특정 빈을 mock 객체로 교체**합니다. 따라서 `@Autowired` 가 동작하여 의존성을 주입할 때 해당 mock 객체가 주입됩니다.

```java
@SpringBootTest
class ThemeServiceTest {

    @MockBean
    private ThemeRepository themeRepository;  // 1. @MockBean으로 mock 객체를 애플리케이션 컨텍스트에 등록
    
    @Autowired
    private ThemeService themeService;  // 2. 의존성 주입 시 실제 ThemeRepository 객체 대신 위에서 등록한 mock 객체가 주입됨
    // ...
}
```

`@Mock` 을 사용하면 스프링 애플리케이션 컨텍스트를 로딩하지 않고도 사용할 수 있기 때문에 단위 테스트 시에 유용하게 사용할 수 있습니다.
반면 `@MockBean` 은 스프링 애플리케이션 컨텍스트를 로딩해야 사용할 수 있기 때문에 스프링 애플리케이션 컨텍스트를 필요로하는 테스트에서 mocking이 필요할 때 유용하게 사용할 수 있습니다.

`@Spy` 와 `@SpyBean` 의 관계 또한 이와 유사합니다. 
단, `@SpyBean` 으로 등록하려는 대상이 인터페이스일 경우, 구현체가 애플리케이션 컨텍스트에 등록되어 있어야 함에 주의해야 합니다.
이는 `@SpyBean` 이 실제 구현된 객체를 감싸는 프록시 형태이기 때문입니다. 애플리케이션 컨텍스트에서 실제 구현체를 일단 꺼내와야 하지만, 등록된 구현체가 없으면 의존성이 해결되지 않아 실패하게 됩니다.

https://cobbybb.tistory.com/16#3.1%20%EB%91%90%20%EB%B2%88%EC%A7%B8%20Refactoring%20%3A%20%40Mock%20%EC%96%B4%EB%85%B8%ED%85%8C%EC%9D%B4%EC%85%98%EC%9C%BC%EB%A1%9C%20%EA%B0%9C%EC%84%A0-1
https://stackoverflow.com/questions/15052984/what-is-the-difference-between-mocking-and-spying-when-using-mockito

**Comment by le2sky** [comment]
Mockito에서 mock과 spy의 차이점은 spy는 실제 객체의 스파이를 생성하여 실제 객체의 메서드를 호출한다는 것입니다. spy 객체의 행동을 따로 지정해주지 않으면 실제 인스턴스의 메서드를 호출합니다.
`@Mock`는 `Mockito.mock()`을 대체할 수 있고, mock 객체를 필드에 바인딩 시켜줍니다.  
반면, `@Spy`는 `Mockito.spy()`을 대체할 수 있고, spy 객체를 필드에 바인딩 시켜줍니다. 
추가로 `@XXBean` 어노테이션은 스프링 영역에 존재하는 어노테이션으로 mock이나 spy를 빈으로 등록하고 필드에 주입해줍니다.

https://cobbybb.tistory.com/16#3.2%20%EC%84%B8%20%EB%B2%88%EC%A7%B8%20Refactoring%20%3A%20%40Mock%2C%20%40InjectMocks%EC%9D%98%20%EC%A1%B0%ED%95%A9-1
